vowpalwabbit/learner.h

```diff
@@ -678,5 +690,212 @@ void multiline_learn_or_predict(multi_learner& base, multi_ex& examples, const u
+  {
+    this->_learner->learn_fd.predict_f = (learn_data::fn)fn_ptr;
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_learn(void (*u)(DataT&, BaseLearnerT&, ExampleT&))
+  {
+    this->_learner->learn_fd.learn_f = (learn_data::fn)u;
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_multipredict(void (*u)(DataT&, BaseLearnerT&, ExampleT&, size_t, size_t, polyprediction*, bool))
+  {
+    this->_learner->learn_fd.multipredict_f = (learn_data::multi_fn)u;
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_update(void (*u)(DataT& data, BaseLearnerT& base, ExampleT&))
+  {
+    this->_learner->learn_fd.update_f = (learn_data::fn)u;
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  // used for active learning and confidence to determine how easily predictions are changed
+  FluentBuilderT& set_sensitivity(float (*u)(DataT& data, base_learner& base, example&))
+  {
+    this->_learner->sensitivity_fd.data = this->_learner->learn_fd.data;
+    this->_learner->sensitivity_fd.sensitivity_f = (sensitivity_data::fn)u;
+
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_learn_returns_prediction(bool learn_returns_prediction)
+  {
+    _learner->learn_returns_prediction = learn_returns_prediction;
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_save_load(void (*sl)(DataT&, io_buf&, bool, bool))
+  {
+    _learner->save_load_fd.save_load_f = (save_load_data::fn)sl;
+    _learner->save_load_fd.data = _learner->learn_fd.data;
+    _learner->save_load_fd.base = _learner->learn_fd.base;
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_finish(void (*f)(DataT&))
+  {
+    _learner->finisher_fd = func_data(_learner->learn_fd.data, _learner->learn_fd.base, (finish_fptr_type)(f));
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_end_pass(void (*f)(DataT&))
+  {
+    _learner->end_pass_fd = func_data(_learner->learn_fd.data, _learner->learn_fd.base, (func_data::fn)f);
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_end_examples(void (*f)(DataT&))
+  {
+    _learner->end_examples_fd = func_data(_learner->learn_fd.data, _learner->learn_fd.base, (func_data::fn)f);
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_init_driver(void (*f)(DataT&))
+  {
+    _learner->init_fd = func_data(_learner->learn_fd.data, _learner->learn_fd.base, (func_data::fn)f);
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_finish_example(void (*f)(vw& all, DataT&, ExampleT&))
+  {
+    _learner->finish_example_fd.data = _learner->learn_fd.data;
+    _learner->finish_example_fd.finish_example_f = (end_fptr_type)(f);
+    return *static_cast<FluentBuilderT*>(this);
+  }
+};
+
+template <class DataT, class ExampleT, class BaseLearnerT>
+struct reduction_learner_builder
```

Comment:
I have never done this myself but I think you can use template types of templates to avoid passing in `DataT`, `ExampleT`, and `BaseLearnerT` twice see [here](https://stackoverflow.com/a/213811)

Reply:
yup I was just wondering if here we could do: `:public common_learner_builder<reduction_learner_builder<DataT, ExampleT, BaseLearnerT>>` and then in `common_learner_builder` struct, do e.g. `FluidBuilderT<DataT>` instead of `DataT`

Reply:
Ah I see what you're saying. However, base_learner_builder and common_learner_builder have a different number of template arguments so the common one can't just apply the 3:
```C++
FluentBuilderT<DataT, ExampleT, BaseLearnerT>
```
 As that would only work for `reduction_learner_builder` and not `base_learner_builder`

Reply:
ah right, never mind then